internal/fxcontext/context.go

```diff
@@ -42,23 +46,24 @@ type Context struct {
 // New always returns Context for use in the service
 func New(ctx gcontext.Context, host service.Host) fx.Context {
 	if host != nil {
-		return &Context{
-			gcontext.WithValue(ctx, _contextLogger, host.Logger()),
-		}
+		ctx = gcontext.WithValue(ctx, _fxContextStore, store{
+			log: host.Logger(),
+		})
 	}
 	return &Context{
 		Context: ctx,
 	}
 }
 
-// Logger returns context based logger
+// Logger returns context based logger. If logger is absent from the context,
+// the function updates the context with a new context based logger
 func (c *Context) Logger() ulog.Log {
-	return c.getLogger()
-}
-
-func (c *Context) getLogger() ulog.Log {
-	if c.Context.Value(_contextLogger) == nil {
-		c.Context = gcontext.WithValue(c.Context, _contextLogger, ulog.Logger())
+	fxctxStore := c.Context.Value(_fxContextStore)
+	if fxctxStore == nil {
+		fxctxStore = store{
+			log: ulog.Logger(),
+		}
```

Comment:
I would add a private `getStore()` function which takes care of replacing the
context with a default store so that `Logger()` is just,

    return c.getStore().log

That way you don't have to repeat this when you add a new property.